File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.cc
#include<iostream>
#include<vector>
#include<string>
#include<queue>
#include<algorithm>

using namespace std;

int map[27][27] = { 0, };
int visit[27][27] = { 0, };

int bfs(int N, int a, int b) {
	queue<pair<int, int>> que;
	pair<int, int> p;
	int idx1[4] = { 0,1,0,-1 };
	int idx2[4] = { 1,0,-1,0 };
	int sol = 0;

	p.first = a;
	p.second = b;
	que.push(p);
	sol++;
	visit[a][b] = 1;
	while (!que.empty()) {
		int a2 = que.front().first;
		int b2 = que.front().second;
		que.pop();

		
		for (int i = 0; i < 4; i++) {
			if (map[a2 + idx1[i]][b2 + idx2[i]] == 1 && visit[a2 + idx1[i]][b2 + idx2[i]] == 0) {
				p.first = a2 + idx1[i];
				p.second = b2 + idx2[i];
				que.push(p);
				sol++;
				visit[p.first][p.second] = 1;
			}
		}

	}
	return sol;
}

int main() {

	vector<int> num;
	int N;
	string str;
	int count = 0;

	cin >> N;

	for (int i = 1; i <= N; i++) {
		cin >> str;
		for (int j = 0; j < str.size(); j++) {
			map[i][j + 1] = str[j] - '0';
		}
	}


	for (int i = 1; i <= N; i++) {
		for (int j = 1; j <= N; j++) {
			if (visit[i][j] == 0 && map[i][j] == 1) {
				num.push_back(bfs(N, i, j));
				count++;
			}
		}
	}

	sort(num.begin(), num.end());

	cout << count << '\n';
	for (int i = 0; i < num.size(); i++) {
		cout << num[i] << '\n';
	}

	return 0;
}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.*;
import java.util.*;

public class Main {

    public static int[][] map;
    public static int[][] check;
    public static int[] m_row = {0, 1, 0, -1};
    public static int[] m_col = {1, 0, -1, 0};
    public static ArrayList<Integer> sol = new ArrayList<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int N = Integer.parseInt(br.readLine());
        map = new int[N+2][N+2];
        check = new int[N+2][N+2];

        // map 세팅
        for (int i = 1; i <= N; i++) {
            String str = br.readLine();
            for (int j=1; j<=str.length(); j++) {
                map[i][j] = str.charAt(j-1) - '0';
            }
        }

        for (int i = 1; i <= N; i++) {
            for (int j = 1; j <= N; j++) {
                if (map[i][j] != 0 && check[i][j] == 0) {
                    sol.add(bfs(i, j));
                }
            }
        }

        Collections.sort(sol);


        bw.write(sol.size() + "\n");
        for (int i = 0; i < sol.size(); i++) {
            bw.write(sol.get(i) + "\n");
        }
        bw.flush();
        bw.close();

    }

    public static int bfs(int row, int col) {
        int count = 0;
        Queue<int[]> que = new LinkedList<>();
        check[row][col] = 1;
        que.offer(new int[] {row, col});

        while (!que.isEmpty()) {
            int[] arr = que.poll();
            count++;

            for (int i = 0; i < 4; i++) {
                if (map[arr[0] + m_row[i]][arr[1] + m_col[i]] != 0 && check[arr[0] + m_row[i]][arr[1] + m_col[i]] == 0) {
                    check[arr[0] + m_row[i]][arr[1] + m_col[i]] = 1;
                    que.offer(new int[] {arr[0] + m_row[i], arr[1] + m_col[i]});
                }
            }
        }

        return count;
    }
}